examples/magika/main.cpp

```diff
@@ -0,0 +1,398 @@
+    const magika_model & model,
+    const int n_files) {
+
+    const auto & hparams = model.hparams;
+
+    static size_t buf_size = ggml_tensor_overhead()*GGML_DEFAULT_GRAPH_SIZE + ggml_graph_overhead();
+    static std::vector<uint8_t> buf(buf_size);
+
+    struct ggml_init_params params = {
+        /*.mem_size   =*/ buf_size,
+        /*.mem_buffer =*/ buf.data(),
+        /*.no_alloc   =*/ true,
+    };
+
+    struct ggml_context * ctx = ggml_init(params);
+
+    struct ggml_cgraph * gf = ggml_new_graph(ctx);
+
+    struct ggml_tensor * input = ggml_new_tensor_3d(ctx, GGML_TYPE_F32, 257, 1536, n_files); // one-hot
+    ggml_set_name(input, "input");
+    ggml_set_input(input);
+
+    struct ggml_tensor * cur;
+
+    // dense
+    cur = ggml_mul_mat(ctx, model.dense_w, input);
+    assert(cur->ne[0] == model.dense_b->ne[0]);
+    cur = ggml_add(ctx, cur, model.dense_b); // [128, 1536, n_files]
+    cur = ggml_gelu(ctx, cur);
+
+    // reshape
+    cur = ggml_reshape_3d(ctx, cur, 512, 384, n_files); // [384, 512, n_files]
+    cur = ggml_cont(ctx, ggml_transpose(ctx, cur));
+
+    // layer normalization
+    cur = ggml_norm(ctx, cur, hparams.f_norm_eps);
+    cur = ggml_mul(ctx, cur, model.layer_norm_gamma); // [384, 512, n_files]
+    cur = ggml_add(ctx, cur, model.layer_norm_beta);  // [384, 512, n_files]
+
+    // dense_1
+    cur = ggml_cont(ctx, ggml_transpose(ctx, cur));
+    cur = ggml_mul_mat(ctx, model.dense_1_w, cur);
+    cur = ggml_add(ctx, cur, model.dense_1_b); // [256, 384, n_files]
+    cur = ggml_gelu(ctx, cur);
+
+    // dense_2
+    cur = ggml_mul_mat(ctx, model.dense_2_w, cur);
+    cur = ggml_add(ctx, cur, model.dense_2_b); // [256, 384, n_files]
+    cur = ggml_gelu(ctx, cur);
+
+    // global_max_pooling1d
+    cur = ggml_cont(ctx, ggml_transpose(ctx, cur)); // [384, 256, n_files]
+
+    // cur = ggml_global_pool_1d(ctx, cur); // [256, n_files]
+    struct ggml_tensor * tmp = ggml_new_tensor_2d(ctx, GGML_TYPE_F32, 256, n_files);
+    cur = ggml_map_custom2_inplace(ctx, tmp, cur, [](struct ggml_tensor * dst, const struct ggml_tensor * t1, const struct ggml_tensor * t2, int ith, int nth, void * user_data) {
+        const char * src_data = (const char *)t2->data;
+        char * dst_data = (char *)dst->data;
+        for (int64_t i3 = 0;  i3 < t2->ne[3]; i3++) {
+            for (int64_t i2 = 0;  i2 < t2->ne[2]; i2++) {
+                for (int64_t i1 = 0;  i1 < t2->ne[1]; i1++) {
+                    float max = -INFINITY;
+                    for (int64_t i0 = 0;  i0 < t2->ne[0]; i0++) {
+                        float val = *(const float *)(src_data + i0*t2->nb[0] + i1*t2->nb[1] + i2*t2->nb[2] + i3*t2->nb[3]);
+                        if (val > max) {
+                            max = val;
+                        }
+                    }
+                    *(float *)(dst_data + i1*dst->nb[0] + i2*dst->nb[1] + i3*dst->nb[2]) = max;
+                }
+            }
+        }
+    }, 1, NULL);
```

Comment:
Does `ggml_pool_1d` not work here?

Reply:
It required a minor change to support 3d tensors, but it seems to work.